Blog Post: Chlorophyll: Synthesis-Aided Compiler for Low-Power Spatial Architectures (#137)
sampsyo/cs6120: content/blog/2019-11-17-chlorophyll/index.md

+++
title = "Chlorophyll: Synthesis-Aided Compiler for Low-Power Spatial Architectures"
extra.author = "Shaojie Xiang and Katy Voor"
extra.bio = """

>   [Shaojie Xiang](https://github.com/Hecmay) is a 2nd year ECE PhD student researching on programming language and distributed system. 

Reviewer: languages (plural) and systems (plural)

>   [Shaojie Xiang](https://github.com/Hecmay) is a 2nd year ECE PhD student researching on programming language and distributed system. 

Reviewer 2: ```suggestion
  [Shaojie Xiang](https://github.com/Hecmay) is a 2nd year ECE PhD student researching on programming languages and distributed systems. 
```

  [Katy Voor](https://github.com/kavoor) is a senior interested in compilers and computer architecture.
"""
+++

===
## Introduction 

> Energy efficiency has been increasingly important for embedded processors and internet of things (IoT) systems. Many spatial hardware architectures have been proposed to address the power consumption problem without compromising the performance. GreenArrays 144 (GA144) is one example of the low power spatial processor. GA144 is a stack-based 18-bit processor without shared memory and fixed clock. To program GA144, users have to write an assembly-like low level programming language [[1]](#1) to define the data storage, data movement and communication logic to make use of the spatial architecture. This paper proposes a programming model and synthesis-aided compiler to compile high level C-like programs into binary executables on GA144. The compiler-optimized program can achieve comparable results with an expert written low level program.

Reviewer 2: ```suggestion
Energy efficiency has been increasingly important for embedded processors and internet of things (IoT) systems. Many spatial hardware architectures have been proposed to address the power consumption problem without compromising the performance. GreenArrays 144 (GA144) is one example of the low power spatial processor. GA144 is a stack-based 18-bit processor without shared memory and fixed clock. To program GA144, users have to write an assembly-like low level programming language [[1]](#1) to define the data storage, data movement and communication logic to make use of the spatial architecture. This paper proposes a programming model and synthesis-aided compiler to compile high level C-like programs into binary executables on GA144. The compiler-optimized program can achieve comparable results with an expert written low level program.

An additional motivation for investigating Chlorophyll is the need for a higher-level abstraction over arrayForth, the language used on GreenArray chips. ArrayForth requires the programmer to partition data structures and code as well as to insert the requisite communication code. Chlorophyll aims to abstract away communication code as well as some partitioning logic that can be inferred by the synthesizer and separator. 
```

### Background 

> Spatial hardware is an architecture where the users are required to assign data, computations and communications explicitly to use its computing resources, storage and interconnect network. Spatial architecture is widely used in domains like High Performance Computing and image processing for its high performance and low power consumption. However, it is especially hard to program and debug due to low programmability and slow compile time. 

Reviewer: I’m actually not sure I agree that spatial architectures are “widely used” in any domain yet. What sort of processors are you referring to? GreenArrays chips, for example, are not exactly common place. I think they probably *will* be, but unless I’m mistaken, we’re not there yet.

Reviewer: It might be nice to briefly explain why this model leads to “low programmability” and “slow compile time.” What is “wrong” with the state of the art?

Reviewer 3: Actually I am not so sure about the definition of spatial architecture, but in my understanding, FPGA is typical spatial architecture and is used as low power accelerators in many domains.

Reviewer: Yeah, good question. It’s not a terribly well defined term, and classifying FPGAs as “just another” spatial architecture is somewhat controversial (although correct IMO). FPGAs are widespread but their use for actual *computation* (as opposed to circuit controls) is somewhat more niche.

Reviewer 3: Sure. This paper is more focused on using spatial architecture as circuit controller. I guess I should rephrase a bit to avoid the potential misunderstanding on this term.

> To tackle the programmability issue, one possible solution is domain specific language (DSL) like Spatial[[2]](#2) and T2S[[3]](#3). DSL provides specialized APIs to help users realize different optimization on target hardware with much less effort. Also people use hardware templates to solve the problem. DNN/FPGA Co-Design[[4]](#4) developed a complication flow from ML workload to spatial accelerators like FPGA and CGRA by searching through template IP pools with certain constraints. Such approaches improve the programmability without sacrificing much flexibility, but still require users to have expertise for the target hardware.

Reviewer: This part of the post has some plurality agreement errors. It might be nice to proofread again with this specifically in mind. For example, in this paragraph:

- “*a* domain specific language”, or “domain specific *languages”
- “A DSL provides”, or “DSLs provide” (not “DSL provide”)
- like FPGAs and CGRAs

> To tackle the programmability issue, one possible solution is domain specific language (DSL) like Spatial[[2]](#2) and T2S[[3]](#3). DSL provides specialized APIs to help users realize different optimization on target hardware with much less effort. Also people use hardware templates to solve the problem. DNN/FPGA Co-Design[[4]](#4) developed a complication flow from ML workload to spatial accelerators like FPGA and CGRA by searching through template IP pools with certain constraints. Such approaches improve the programmability without sacrificing much flexibility, but still require users to have expertise for the target hardware.

Reviewer: I recommend putting spaces before citation brackets, like this [3].

> To tackle the programmability issue, one possible solution is domain specific language (DSL) like Spatial[[2]](#2) and T2S[[3]](#3). DSL provides specialized APIs to help users realize different optimization on target hardware with much less effort. Also people use hardware templates to solve the problem. DNN/FPGA Co-Design[[4]](#4) developed a complication flow from ML workload to spatial accelerators like FPGA and CGRA by searching through template IP pools with certain constraints. Such approaches improve the programmability without sacrificing much flexibility, but still require users to have expertise for the target hardware.

Reviewer 2: ```suggestion
To tackle the programmability issue, one possible solution is a domain specific language (DSL) like Spatial[[2]](#2) and T2S[[3]](#3). DSLs provide specialized APIs to help users realize different optimization on target hardware with much less effort. People can also use hardware templates to solve the problem. DNN/FPGA Co-Design [[4]](#4) developed a complication flow from an ML workload to spatial accelerators like FPGAs and CGRAs by searching through template IP pools with certain constraints. Such approaches improve the programmability without sacrificing much flexibility, but still require users to have expertise for the target hardware.
```

> To tackle the programmability issue, one possible solution is domain specific language (DSL) like Spatial[[2]](#2) and T2S[[3]](#3). DSL provides specialized APIs to help users realize different optimization on target hardware with much less effort. Also people use hardware templates to solve the problem. DNN/FPGA Co-Design[[4]](#4) developed a complication flow from ML workload to spatial accelerators like FPGA and CGRA by searching through template IP pools with certain constraints. Such approaches improve the programmability without sacrificing much flexibility, but still require users to have expertise for the target hardware.

Reviewer 2: ```suggestion
To tackle the programmability issue, one possible solution is domain specific language (DSL) like Spatial [[2]](#2) and T2S [[3]](#3). DSL provides specialized APIs to help users realize different optimization on target hardware with much less effort. Also people use hardware templates to solve the problem. DNN/FPGA Co-Design [[4]](#4) developed a complication flow from ML workload to spatial accelerators like FPGA and CGRA by searching through template IP pools with certain constraints. Such approaches improve the programmability without sacrificing much flexibility, but still require users to have expertise for the target hardware.
```

### Chlorophyll Overview

Chlorophyll decomposes the problem of compiling high level programs into spatial machine code into four sub-problems:

<img src="flow.png" width="700" >

> * **Partitioning**: search the partitioning scheme that minimizes the communication between different logical cores

Reviewer 2: ```suggestion
* **Partitioning**: Search the partitioning scheme that minimizes the communication between different logical cores
```

> * **Layout**: assign the partitioned program segments to physical cores such that the actual communication cost is minimized 

Reviewer 2: ```suggestion
* **Layout**: Assign the partitioned program segments to physical cores such that the actual communication cost is minimized 
```

> * **Code Separation**: generate communication code for data movements between physical cores

Reviewer 2: ```suggestion
* **Code Separation**: Generate communication code for data movements between physical cores
```

> * **Code Generation**: Search and generate high performance binary code taking advantage of hardware features of target machine

Reviewer: This sentence is capitalized, while the other bullet points are not.

> * **Code Generation**: Search and generate high performance binary code taking advantage of hardware features of target machine

Reviewer: of *the* target machine

> * **Code Generation**: Search and generate high performance binary code taking advantage of hardware features of target machine

Reviewer 2: ```suggestion
* **Code Generation**: Search and generate high performance binary code taking advantage of hardware features of the target machine
```

> The users can develop their programs in C and compile to GrennArray 144 without worrying about the communication and data movement. The compiler partitions the code, explores the design space, and automatically optimizes the program to improve the performance.

Reviewer: “GrennArray”

> The users can develop their programs in C and compile to GrennArray 144 without worrying about the communication and data movement. The compiler partitions the code, explores the design space, and automatically optimizes the program to improve the performance.

Reviewer 2: ```suggestion
The users can develop their programs in C and compile to GA144 without worrying about the communication and data movement. The compiler partitions the code, explores the design space, and automatically optimizes the program to improve the performance.
```

### Partitioning Synthesizer

#### Type System

Chlorophyll provides an annotation based syntax for users to indicate partition types for variables, arrays and operators. Partition types can be declared using annotation or inferred by the synthesizer. Here follows an example of annotating arrays into different partitions.

```cpp
int@{[0:32]=0,[32:64]=1} x[64]; 
int@{[0:32]=2,[32:64]=3} y[64]; 
int@{[0:32]=4,[32:64]=5} z[64]; 
int@6 r = 0;
for (i from 0 to 64) {
  z[i] = leftrotate(x[i],y[i],r)!4 -@place(z[i]) 1; 
  if (r >@6 32) r = 0;
}
```

`@` in the program is an indicator of the partition index, and `@place` will return the partition type the argument. `!` is an operation of sending data from one partition to another. In the example above, array x, y and z are partitioned into two logical cores (first half and second half stored in separate cores). The computation of `leftrotate` function may happen in the same or a different logical core depending on the result of type inference. 

> For loops in Chlorophyll programs, the synthesizer conducts loop splitting to make variables in each splitted loop have the same partition ID. The loop in the example above will be decomposed into multiple non-overlapping sub-loops.

Reviewer: splitted -> split

> For loops in Chlorophyll programs, the synthesizer conducts loop splitting to make variables in each splitted loop have the same partition ID. The loop in the example above will be decomposed into multiple non-overlapping sub-loops.

Reviewer 2: ```suggestion
For loops in Chlorophyll programs, the synthesizer conducts loop splitting to make variables in each split loop have the same partition ID. The loop in the example above will be decomposed into multiple non-overlapping sub-loops.
```

```cpp
for (i from 0 to 32) {
  z[i] = leftrotate(x[i]!8,y[i]!8,r!9)!4 -@4 1;
  r = r +@6 1;
  if (r >@6 32) r = 0;
}
for (i from 32 to 64) {
  z[i] = leftrotate(x[i]!8,y[i]!8,r!9)!5 -@5 1;
  r = r +@6 1;
  if (r >@6 32) r = 0;
}
```

In the partitioning stage, Chlorophyll compiler performs type checking and type inference to generate the completely partitioned code, where each operand will have an explicit partition type as well as the destination partition ID (if it will be used by operations on other logical cores). The partitioning is conducted with the following two key functions:

* **Communication Interpreter**: The interpreter calculates the communication count of given program `P` with every valid input `x` under a complete partition type annotation $\sigma$. The maximum communication count is defined as
$$
MaxComm(P, \sigma) = max_{x}Comm(P, \sigma, x)

> $$ If the input program is partially annotated, the interpreter returns a formula in terms of symbolic variables for partition types. The returned value is used as an input to a constraint-based solver to search for an optimal partitioning scheme for the given program (i.e. better type inference on the un-annotated variables). 

Reviewer: Use a comma after i.e. https://capra.cs.cornell.edu/styleguide/#egie

> $$ If the input program is partially annotated, the interpreter returns a formula in terms of symbolic variables for partition types. The returned value is used as an input to a constraint-based solver to search for an optimal partitioning scheme for the given program (i.e. better type inference on the un-annotated variables). 

Reviewer 2: ```suggestion
$$ If the input program is partially annotated, the interpreter returns a formula in terms of symbolic variables for partition types. The returned value is used as an input to a constraint-based solver to search for an optimal partitioning scheme for the given program (i.e., better type inference on the un-annotated variables). 
```

> $$ If the input program is partially annotated, the interpreter returns a formula in terms of symbolic variables for partition types. The returned value is used as an input to a constraint-based solver to search for an optimal partitioning scheme for the given program (i.e. better type inference on the un-annotated variables). 

Reviewer 2: ```suggestion
$$ If the input program is partially annotated, the interpreter returns a formula in terms of symbolic variables for partition types. The returned value is used as an input to a constraint-based solver to search for an optimal partitioning scheme for the given program (i.e., better type inference on the un-annotated variables). 
```

* **Partition Space Checker**: Given a complete partition scheme for each variable, array and other constructs in the program, the space checker checks the space usage of in each partition, to make sure that the memory usage is no larger than the storage capacity on the core.

The partitioning synthesizer utilizes these two functions as constraints to a backend solver to infer the prtitioning scheme minimizing the communication while not violating the memory constraint. Every time the solver finds a feasible solution, the synthesizer will lower the upper bound constraint of communication count to push for a better solution in next runs.

### Layout 

The layout problem is formalized as a quadratic assignment problem (QAP), where we search for the assignment function to minimize the total communication cost:
* facilities set $P$ (a set of partition types we derived from partitioning synthesizer)

> * the weight distance $W(a,b)$ returning the weight factor between two facilities $a$ and $b$. 

Reviewer: Some of these bullet points have periods at the end and others do not.

> * the weight distance $W(a,b)$ returning the weight factor between two facilities $a$ and $b$. 

Reviewer 2: ```suggestion
* the weight distance $W(a,b)$ returning the weight factor between two facilities $a$ and $b$
```

* function $D(a, b)$ returning the distance between two given locations

> * Assignment function mapping different facility (in our case function) into a specific location (i.e. logical core on GA144). 

Reviewer 2: ```suggestion
* Assignment function mapping different facility (in our case function) into a specific location (i.e., logical core on GA144)
```

Finally, with both results written to the data stack of the core mapped by partition 3, the operation `(3+7)` occurs by reading its data stack to source operands.

This process is much more challenging with control flow and functions. When considering functions, a function call in the original program translates to a function call at each of the cores on which the function resides.

<!-- TODO maybe talk about arrays -->

### Code Generation
<img src="overview.png" width="600" >
Typically, code generation is performed using a dynamic programming algorithm that utilizes local optimization to optimize larger and larger code fragments. Think of tiling an abstract assembly tree and building up larger and larger optimized tiles from smaller ones. This approach is not as effective for code generation for nontraditional architectures with nonlocal optimizations, such as optimizations for circular stacks. 

Instead of writing new transformation rules, the authors search for an optimized program in the space of candidate programs. One way to do this is through superoptimization.

> *Superoptimization* is the process of searching the space of all instruction sequences and verifying the candidate programs against a reference implementation with naively-generated code. Unfortunately, according to this paper, superoptimization only scales to roughly 25 instructions. However, there has been more recent work to mitigate this with increased contextual information and early pruning of invalid candidate programs.[[4]](#4)

Reviewer: Citations go before the period, like this [4].

> *Superoptimization* is the process of searching the space of all instruction sequences and verifying the candidate programs against a reference implementation with naively-generated code. Unfortunately, according to this paper, superoptimization only scales to roughly 25 instructions. However, there has been more recent work to mitigate this with increased contextual information and early pruning of invalid candidate programs.[[4]](#4)

Reviewer 2: ```suggestion
*Superoptimization* is the process of searching the space of all instruction sequences and verifying the candidate programs against a reference implementation with naively-generated code. Unfortunately, according to this paper, superoptimization only scales to roughly 25 instructions. However, there has been more recent work to mitigate this with increased contextual information and early pruning of invalid candidate programs [[4]](#4).
```

This engenders additional optimization as well. Consider the corresponding stacks for a subtract operation.

<img src="subexample.png" width="400" >

Here, b and a are popped from the stack, subtracted, and the result is pushed back onto the stack. The *P* that results in only `b-a` left on the stack is 8 instructions long, with 3 instructions to remove the remaining garbage value `a` from the stack. However, as noted before, it is in fact legal to leave `a` at the bottom of the stack, thus saving 3 instructions. 

#### Sliding Windows
To scale superoptimization to real world programs, the authors employ the *sliding windows technique*.
This algorithm merges superoptimizable units into a superoptimizable segment. It starts with an empty segment, superoptimizes the largest sequence possible from the input instruction sequence, and appends the result to the output, if valid.
In the case where no valid segment is found from the superoptimizer, it appends the the first unit from the segment to the output, and attempts to build another sequence to superoptimize. 
Then, it iterates until the instruction sequence is empty.

> The superoptimizer executes a binary search over programs given a cost model. It uses counterexample-guided inductive synthesis (CEGIS) to synthesize a program of cost k, and if one exists, to synthesize a program of cost k/2, and so on.

Reviewer: Maybe use $ for math around k and k/2?

> The superoptimizer executes a binary search over programs given a cost model. It uses counterexample-guided inductive synthesis (CEGIS) to synthesize a program of cost k, and if one exists, to synthesize a program of cost k/2, and so on.

Reviewer 2: ```suggestion
The superoptimizer executes a binary search over programs given a cost model. It uses counterexample-guided inductive synthesis (CEGIS) to synthesize a program of cost $k$, and if one exists, to synthesize a program of cost $k/2$, and so on.
```

> When perusing papers that cite this one, we found the primary critique of the *sliding windows technique* to be the focus on local optimization and inability to make whole-program optimality guarantees. Other synthesis approaches such as Metasketches, claim to provide these guarantees.[[6]](#6)

Reviewer: Citations before periods, as above. (Also, I don’t think you need to capitalize metasketches.)

> When perusing papers that cite this one, we found the primary critique of the *sliding windows technique* to be the focus on local optimization and inability to make whole-program optimality guarantees. Other synthesis approaches such as Metasketches, claim to provide these guarantees.[[6]](#6)

Reviewer 2: ```suggestion
When perusing papers that cite this one, we found the primary critique of the *sliding windows technique* to be the focus on local optimization and inability to make whole-program optimality guarantees. Other synthesis approaches such as Metasketches, claim to provide these guarantees [[6]](#6).
```

### Opportunity for Improvement with Human Insight
One last item to note is the time it takes for the superoptimizer to converge on a valid optimized program is significant. Below we see the results for single-core and multicore benchmarks with their corresponding convergence times.

<img src="singlecore.png" width="500" >
<img src="multicore.png" width="500">

> The authors' claim that injecting human insight into the superoptimizer in the form of templates or pinning code to cores can help mitigate these times. They provide minimal examples and evidence of the potential impact of this human insight. We note the tradeoff between convergence times and programmability, one of their primary motivations for Chlorophyll as a higher-level abstraction over arrayForth.

Reviewer: I don’t think you’ve mentioned arrayForth yet, so maybe briefly introduce what that is?

### Optimization Opportunity Loss

> The separation of partitioning and superoptimization engenders a few places for improvement. Chlorophyll uses a schedule-oblivious routing strategy. Therefore, if core A can communicate with core B with core X or core Y, it will choose the route arbitrarily. If core X is busy during this time, the optimal route would be through core B, instead of waiting on X to finish its work. When determining the importance of this scheduling issue, we wanted to know what proportion of execution time is dedicated to communication cost, including time spent waiting. Given the strict space constraints, it is unclear whether inserting instructions to conditionally use different communication instructions saves enough execution time to be worth the space. Further benchmarks would be required to determine if it is worth it, and if so, in which cases. 

Reviewer: In “core A can communicate with core B with core X or core Y,” with -> via?

> The separation of partitioning and superoptimization engenders a few places for improvement. Chlorophyll uses a schedule-oblivious routing strategy. Therefore, if core A can communicate with core B with core X or core Y, it will choose the route arbitrarily. If core X is busy during this time, the optimal route would be through core B, instead of waiting on X to finish its work. When determining the importance of this scheduling issue, we wanted to know what proportion of execution time is dedicated to communication cost, including time spent waiting. Given the strict space constraints, it is unclear whether inserting instructions to conditionally use different communication instructions saves enough execution time to be worth the space. Further benchmarks would be required to determine if it is worth it, and if so, in which cases. 

Reviewer 2: ```suggestion
The separation of partitioning and superoptimization engenders a few places for improvement. Chlorophyll uses a schedule-oblivious routing strategy. Therefore, if core A can communicate with core B via core X or core Y, it will choose the route arbitrarily. If core X is busy during this time, the optimal route would be through core B, instead of waiting on X to finish its work. When determining the importance of this scheduling issue, we wanted to know what proportion of execution time is dedicated to communication cost, including time spent waiting. Given the strict space constraints, it is unclear whether inserting instructions to conditionally use different communication instructions saves enough execution time to be worth the space. Further benchmarks would be required to determine if it is worth it, and if so, in which cases. 
```

An additional motivation for investigating Chlorophyll is the need for a higher-level abstraction over arrayForth, the language used on GreenArray chips. ArrayForth requires the programmer to partition data structures and code as well as to insert the requisite communication code. Chlorophyll aims to abstract away communication code as well as some partitioning logic that can be inferred by the synthesizer and separator. 

### Background 

> Spatial hardware is an architecture where the users are required to assign data, computations and communications explicitly to use its computing resources, storage and interconnect network. Compared with general purpose processors, spatial architecture can achieve comparatively good, or even better performance as well as high energy efficiency for some specific applications. However, spatial hardware is especially hard to program and debug cause users have to manually manage the low level hardware resources and efficiently transform their program in order to take advantage of the hardware features. 

Reviewer 2: ```suggestion
Spatial hardware is an architecture where the users are required to assign data, computations, and communications explicitly to use its computing resources, storage, and interconnect network. Compared with general purpose processors, spatial architectures can achieve comparatively good, or even better performance and energy efficiency for some specific applications. However, spatial hardware is especially hard to program and debug as users have to manually manage the low level hardware resources and efficiently transform their programs in order to take advantage of the hardware features. 
```
